power_grid_model_c/power_grid_model/include/power_grid_model/sparse_mapping.hpp

```diff
@@ -50,35 +50,164 @@ struct SparseMapping {
         ++counter[entry.second];
     }
     // accumulate value to indptr
+    // [a, b, c, d], the inclusive scan would produce [a, a+b, a+b+c, a+b+c+d]
+    // counter = {0, 2, 2, 1, 0, 1, 0}, indptr becomes {0, 0, 2, 4, 5, 5, 6, 6}
     std::inclusive_scan(counter.cbegin(), counter.cend(), sparse_mapping.indptr.begin() + 1);
     // copy back
+    // copy everything except the indptr[0] to counter.
+    // counter becomes {0, 2, 4, 5, 5, 6, 6}
     std::copy(sparse_mapping.indptr.cbegin() + 1, sparse_mapping.indptr.cend(), counter.begin());
     // sort
+    // crbegin() points to the element at the end of the container
+    // crend() points to the position before the first element
+    // at this point: entries[{0,3},{1,5},{2,2},{3,1},{4,1},{5,2}], counter {0, 2, 4, 5, 5, 6, 6}
+    // Reverse order the first entry in entries: {5,2}, it_entry->second = 2
+    // counter[2] is 4, decremented to 3(counter becomes {0, 2, 3, 5, 5, 6, 6}), thus sparse_mapping.reorder[3] is set
+    // to it_entry->first which is 5. sparse_mapping.reorder = {0, 0, 0, 5, 0, 0},
+    // Final result: sparse_mapping.reorder = {3, 4, 2, 5, 0, 1}
     for (auto it_entry = entries.crbegin(); it_entry != entries.crend(); ++it_entry) {
         sparse_mapping.reorder[--counter[it_entry->second]] = it_entry->first;
     }
+    // the first is the index before reordering of load_gen
+    // it entry second is the index of the bus
+    // reorder at the end contains for each new index of the load gen the old index of the load gen in a linear way
     assert(sparse_mapping.indptr[0] == 0);
     assert(sparse_mapping.indptr.back() == n_A);
     return sparse_mapping;
 }
 
-} // namespace power_grid_model
+struct DenseMapping {
+    IdxVector indvector;
+    IdxVector reorder;
+};
+
+inline DenseMapping build_dense_mapping(IdxVector const& idx_B_in_A, Idx const n_B) {
+    auto const n_A = static_cast<Idx>(idx_B_in_A.size());
+    auto const r = std::max_element(idx_B_in_A.begin(), idx_B_in_A.end())[0];
+    // for counting sort nr_of_operations = 1 + 3*n_A + 2*n_B + 2*n_A + 2*n_B - 1 + 1 + n_A + 1 + 14 * n_A
+    // nr_of_operations = 20*n_A + 4*n_B + 2
+    auto const c = 20 * n_A + 4 * n_B + 2;
+    // for n log n sort nr of operations = 1 + 7*n_A + n_A*log(n_A)?
+    using DenseEntry = std::pair<Idx, Idx>;
+    DenseMapping dense_mapping;
+
+    if (n_A * log(n_A) < c * r) {
+        // 1 operation?
+        std::vector<DenseEntry> mapping_to_from;
+
+        // n_A operations
+        auto index_range = std::ranges::views::iota(0, static_cast<Idx>(idx_B_in_A.size()));
+
+        // n_A * 2 operations?
+        std::ranges::transform(idx_B_in_A, index_range, std::back_inserter(mapping_to_from),
+                               [](Idx value, Idx orig_idx) {
+                                   return std::pair{value, orig_idx};
+                               });
+        // n_A log n_A?
+        std::ranges::sort(mapping_to_from); // runs in O(n log n)
```

Comment:
you'll have to measure the prefactor before `n log n` by providing worst-case scenarios to the algorithm for different values of `n_A`, e.g. `100, 1000, 10000, 100000, 1000000`)

(please refer to wikipedia; in particular `introsort` (which is used by the GNU STL and MSVC STL implementations, as mentioned by https://en.cppreference.com/w/cpp/algorithm/ranges/sort , as well as the LLVM STL as mentioned by wikipedia. the constant term is negligible for large `n_A` which are interesting for us)